Zephyr port to imx8mnano

waseemh40 opened the pull request:
Hello,

As per instructions, I have modified the branch/repo. It provides basic operations support i.e. UART and GPIO. I have also added and tested a sample application. Finally, this is a multi-commit version, with each commit clearly explain it's purpose/modifications. Hope it is OK now. I am not sure if hal_nxp has been already added or do I have to create a pull request for that also?

Regards,

Waseem

waseemh40 committed:
soc: Added iMx8MNano SoC related files.

Signed-off-by: Waseem Hassan <waseemh40@gmail.com>

waseemh40 committed:
boards: Added support for Digi CC8MNano DVK.

Signed-off-by: Waseem Hassan <waseemh40@gmail.com>

waseemh40 committed:
dts: Added device tree source for iMx8MNano SoC. Also, modified CC8MNano DVK for inclusion of gpio2.

Signed-off-by: Waseem Hassan <waseemh40@gmail.com>

waseemh40 committed:
drivers/clock_control: Added clock support for iMx8MNano SoC.

Signed-off-by: Waseem Hassan <waseemh40@gmail.com>

waseemh40 committed:
drivers: modified mcux_igpio driver for iMx8MNano GPIO support. soc: Modified Kconfig options. samples: added a simple thread based demo application for testing UART4 and GPIO2.

Signed-off-by: Waseem Hassan <waseemh40@gmail.com>

galak commented on boards/arm/cc8mnano_dvk/doc/img/index.rst:
```diff
@@ -0,0 +1,47 @@
+Digi CC8MNANO DVK
+###############
+
+Digi ConnectCore® 8M Nano SPECIFICATIONS 					
```
This looks like it needs some formatting fixing.

galak commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -108,6 +108,9 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
 				(CLOCK_GetRootPostDivider(kCLOCK_RootUart4)) /
 				10;
+		if(CONFIG_SOC_SERIES_IMX8MN_M7){
+			*rate=24000000;			//24MHz oscillator is used for CC8MNano DVK
```
Is the rate always 24Mhz for IMX8 nano?

waseemh40 replied:
Yes. We ported Zephyr for a low power use case scenario. Therefore, we used a clock source which:

1. Saves power due to lower clock rate.
2. Runs independently even when Linux/A53 cores are sleeping.

galak commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -47,11 +47,21 @@ static int mcux_igpio_configure(const struct device *dev,
+#else
+        base->DR |= BIT(pin); /* Set pin output to low level.*/
+#endif
+//		base->DR_SET = BIT(pin);
```
remove the commented out line.

waseemh40 replied:
OK

galak commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -47,11 +47,21 @@ static int mcux_igpio_configure(const struct device *dev,
+#else
+        base->DR &= ~BIT(pin); /* Clear pin output to low level.*/
+#endif	
+//		base->DR_CLEAR = BIT(pin);
```
removed commented out line.

waseemh40 replied:
OK

galak commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -86,8 +96,12 @@ static int mcux_igpio_port_set_bits_raw(const struct device *dev,
+#else
+        base->DR |= mask; /* Set mask*/
+#endif
+//	base->DR_SET = mask;
```
removed commented out line.

waseemh40 replied:
OK

galak commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -97,8 +111,12 @@ static int mcux_igpio_port_clear_bits_raw(const struct device *dev,
+#else
+        base->DR &= ~mask; /* Clear mask*/
+#endif	
+//	base->DR_CLEAR = mask;
```
removed commented out line.

waseemh40 replied:
OK

galak commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -243,3 +265,4 @@ static const struct gpio_driver_api mcux_igpio_driver_api = {
 	}
 
 DT_INST_FOREACH_STATUS_OKAY(MCUX_IGPIO_INIT)
+
```
remove extra blank line here.

galak commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -108,8 +126,12 @@ static int mcux_igpio_port_toggle_bits(const struct device *dev,
 	GPIO_Type *base = config->base;
-
-	base->DR_TOGGLE = mask;
+#if (defined(FSL_FEATURE_IGPIO_HAS_DR_CLEAR) && FSL_FEATURE_IGPIO_HAS_DR_CLEAR)
```
should probably be FSL_FEATURE_IGPIO_HAS_DR_TOGGLE instead of CLEAR

galak commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -86,8 +96,12 @@ static int mcux_igpio_port_set_bits_raw(const struct device *dev,
 	GPIO_Type *base = config->base;
-
-	base->DR_SET = mask;
+#if (defined(FSL_FEATURE_IGPIO_HAS_DR_CLEAR) && FSL_FEATURE_IGPIO_HAS_DR_CLEAR)
```
Should probably be FSL_FEATURE_IGPIO_HAS_DR_SET instead of CLEAR.

galak commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -47,11 +47,21 @@ static int mcux_igpio_configure(const struct device *dev,
 
 	if (flags & GPIO_OUTPUT_INIT_HIGH) {
-		base->DR_SET = BIT(pin);
+#if (defined(FSL_FEATURE_IGPIO_HAS_DR_CLEAR) && FSL_FEATURE_IGPIO_HAS_DR_CLEAR)
```
Should probably be FSL_FEATURE_IGPIO_HAS_DR_SET instead of CLEAR.

galak replied:
I think you missed one.

galak commented on soc/arm/nxp_imx/mimx8mn6_m7/soc.c:
```diff
@@ -0,0 +1,238 @@
+/* OSC/PLL is already initialized by ROM and Cortex-A53 (u-boot) */
+static void SOC_RdcInit(void)
+{
+    /* Move M7 core to specific RDC domain 1 */
```
fix whitespace issues here, should be using tabs instead of spaces.

galak reviewed (changes requested):
Looks pretty good, thanks for resubmitting this and keeping it smaller.

* some comments on the GPIO driver change
* need to fix whitespace/formatting in a few places
* I'd suggest removing samples/threads_based_uart_led as I dont think this adds anything other samples don't already cover.

waseemh40 committed:
soc: format fixing imx8mnano soc.c. drivers: format fixing in mcux_igpio driver. boards: formatting board documentation.

Signed-off-by: Waseem Hassan <waseemh40@gmail.com>

waseemh40 commented:
Added the requested changes/formatting.

-Waseem

waseemh40 committed:
samples: removed threads_based_uart_led demo

Signed-off-by: Waseem Hassan <waseemh40@gmail.com>

waseemh40 committed:
boards: fixed documenation. boards: drivers: soc: dts: formatted code.

Signed-off-by: Waseem Hassan <waseemh40@gmail.com>

waseemh40 commented:
Hi,

@galak I was wondering what happens next? I have added the required changes. 

Thanks in advance.

-Waseem

waseemh40 commented:
@galak Should I also merge all the commits here into one? or maybe merge only last 5 commits, such that there are only 4 commits i.e. 
1. soc: Added iMx8MNano SoC related files.
2. boards: Added support for Digi CC8MNano DVK.
3. dts: Added device tree source for iMx8MNano SoC. Also, modified CC8MNano DVK for inclusion of gpio2.
4. drivers: Added clock and gpio driver for iMx8MNano SoC. Formatted documentation.

arvid-r commented:
This looks great! I hope it can be merged 👍

waseemh40 commented:
Hi,

@galak It has been a long time. Could you kindly look into it and maybe merge th pull if things are OK?

-Waseem

waseemh40 commented:
Hi,

@MaureenHelm I contacted Kumar Gala for further comments/review but he has replied. Could you kindly look into it and maybe merge if things are OK?

-Waseem

MaureenHelm commented:
> @MaureenHelm I contacted Kumar Gala for further comments/review but he has replied. Could you kindly look into it and maybe merge if things are OK?

@galak isn't actively working on zephyr at the moment. @dleach02  will you please pick this up?

This PR hasn't been rebased in 8 months. Will you please rebase to the latest main?

waseemh40 commented:
Tried rebasing it to the updated main. Everything else remains same. Can you kindly check now?

MaureenHelm commented:
> Tried rebasing it to the updated main. Everything else remains same. Can you kindly check now?

Something went wrong in your rebase. This PR is now showing over 9000 commits

waseemh40 commented:
Now it is 8 commits ahead of the zephyrproject:main branch. Used force push to resolve the 9k commits issue. Fingers crossed :-)

-Waseem

MaureenHelm commented:
> @MaureenHelm Now it is 8 commits ahead of the zephyrproject:main branch. Used force push to resolve the 9k commits issue. Fingers crossed :-)

Looks better, but there are still CI failures that you need to resolve.

povergoing commented on soc/arm/nxp_imx/mimx8mn6_m7/soc.c:
```diff
@@ -0,0 +1,226 @@
+	 */
+
+	/* Region 0 [0x0000_0000 - 0x4000_0000] : Memory with Device type, not executable, not shareable, non-cacheable. */
+	MPU->RBAR = ARM_MPU_RBAR(0, 0x00000000U);
```
Is it possible to do the mpu config like what https://github.com/zephyrproject-rtos/zephyr/blob/feffe639a3f8a2d2459d21a390ffd2d9d96c2437/soc/arm/nxp_imx/mimx8ml8_m7/mpu_regions.c#L21 does?

waseemh40 commented:
@MaureenHelm Should I try to fix ALL the errors or just x6 errors in the failed check "Compliance Checks / Run compliance checks on patch series (PR) (pull_request) ". These errors are related with coding style/white-spaces, whereas errors under Unit Test Results, twister-build (3), twister-build (2) and twister-build (1) checks are due to missing device drivers (e.g. fsl_gpc.h) which comes after adding/fixing hal_nxp part, I guess.
 
@povergoing I am using soc.h based on NXPs SDK without modifying too many things. The same way it is done for mimx8mm6_m4... Hope it is OK :-)

waseemh40 committed:
boards, dts, soc: fixed code formatting style.

Signed-off-by: Waseem Hassan <waseemh40@gmail.com>

waseemh40 commented:
@MaureenHelm Hi again, I have fixed the errors marked required (x6) as explained in the previous message. Kindly let me know the if this part is OK. I have noticed that hal_nxp has changed a lot since last year. Therefore, I guess I have to work with that part as well in order to add support for 8M Nano.

-Waseem

MaureenHelm commented:
> Should I try to fix ALL the errors or just x6 errors in the failed check "Compliance Checks / Run compliance checks on patch series (PR) (pull_request) ". These errors are related with coding style/white-spaces, whereas errors under Unit Test Results, twister-build (3), twister-build (2) and twister-build (1) checks are due to missing device drivers (e.g. fsl_gpc.h) which comes after adding/fixing hal_nxp part, I guess.

All of the CI failures need to be fixed. You also need to rebase and resolve a merge conflict.

dleach02 commented:
@waseemh40 will you be returning to this PR?
